ngsPETSc/utils/firedrake/meshes.py

```diff
@@ -70,105 +76,156 @@ def refineMarkedElements(self, mark):
-            #Bend if not shared
-            if np.sum(shared) == 1:
-                if isInMesh:
-                    p = [np.argmin(np.sum((pts - pt)**2, axis=1))
-                            for pt in V[cellMap.values[Idx]][0:refPts.shape[0]]]
-                    curvedPhysPts[i] = curvedPhysPts[i][p]
-                    res = np.linalg.norm(pts[p]-V[cellMap.values[Idx]][0:refPts.shape[0]])
-                    if res > tol:
-                        fd.logging.warning("[{}] Not able to curve Firedrake element {} \
-                            ({}) -- residual: {}".format(self.comm.rank, Idx,i, res))
-                    else:
-                        for j, datIdx in enumerate(cellMap.values[Idx][0:refPts.shape[0]]):
-                            for dim in range(self.geometric_dimension()):
-                                coo = curvedPhysPts[i][j][dim]
-                                newFunctionCoordinates.sub(dim).dat.data[datIdx] = coo
-            else:
-                if isInMesh:
-                    p = [np.argmin(np.sum((pts - pt)**2, axis=1))
-                            for pt in V[cellMap.values[Idx]][0:refPts.shape[0]]]
-                    curvedPhysPts[i] = curvedPhysPts[i][p]
-                    res = np.linalg.norm(pts[p]-V[cellMap.values[Idx]][0:refPts.shape[0]])
-                else:
-                    res = np.inf
-                res = self.comm.gather(res, root=0)
-                res = self.comm.bcast(res, root=0)
-                rank = np.argmin(res)
-                if self.comm.rank == rank:
-                    if res[rank] > tol:
-                        fd.logging.warning("[{}, {}] Not able to curve Firedrake element {} \
-                            ({}) -- residual: {}".format(self.comm.rank, shared, Idx,i, res))
-                    else:
-                        for j, datIdx in enumerate(cellMap.values[Idx][0:refPts.shape[0]]):
-                            for dim in range(self.geometric_dimension()):
-                                coo = curvedPhysPts[i][j][dim]
-                                newFunctionCoordinates.sub(dim).dat.data[datIdx] = coo
-
-    return newFunctionCoordinates
+        curved = self.comm.bcast(None, root=0)
+        # Construct numpy arrays as buffers to receive physical domain data
+        ncurved = np.sum(curved)
+        physical_space_points = np.zeros(
+            (ncurved, reference_space_points.shape[0], geom_dim)
+        )
+        curved_space_points = np.zeros(
+            (ncurved, reference_space_points.shape[0], geom_dim)
+        )
+
+    # Broadcast curved cell point data
+    self.comm.Bcast(physical_space_points, root=0)
+    self.comm.Bcast(curved_space_points, root=0)
+    cell_node_map = new_coordinates.cell_node_map()
+
+    # Select only the points in curved cells
+    barycentres = np.average(physical_space_points, axis=1)
+    ng_index = [*map(lambda x: self.locate_cell(x, tolerance=0.01), barycentres)]
```

Comment:
The default tolerance is 0.5, this is a relative for `locate_cell` rather than absolute tolerance. We already have to pass a tolerance to this call, do we want to add another?

Reply:
The two tolerance would have two different meanings no? One is the tolerance for snapping, and the other is for finding the cell to which the barycenter belong to.

Reply:
Hmm, currently this function is underdocumented...

Reply:
I documented the two :)  Let's see if tests pass or if we are explicitly passing to somewhere (if so I'll fix this later).

Reply:
What about CG? I think that should be a bit more explicit as a keyword and lowercase. But regardless it should be documented.

Reply:
I've implemented you suggestions :)

Reply:
Ok, Firedrake tests are failing. I'll open a PR for the Firedrake user meeting next Wednesday and then merge in this PR :) Sounds good?

Reply:
So it turns out those errors are coming from ngsPETSc!!! Commits incoming...

Reply:
Looks like it's all fixed now